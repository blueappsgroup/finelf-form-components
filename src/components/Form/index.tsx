import React, { FC, ReactElement, useMemo, useState } from 'react'
import { Formik, Form } from 'formik'
import styled from 'styled-components'

import { ThemeProvider } from '../../consts/theme'
import { device } from '../../consts/sizes'
import { FormProps, FormValuesType } from '../../types'
import RedirectPage from '../RedirectPage'
import {
  FormContext,
  getFormValuesFromCache,
  resetFormValueCache,
  handleSendDataToApi,
  getFieldsValuesFromUrl,
} from '../../utils'
import { formStatuses } from '../../consts/form'

const StyledForm = styled(Form)`
  display: flex;
  flex-direction: column;
  max-width: ${(props): string => props.theme.formMaxWidth};
  justify-self: center;
  background: ${(props): string => props.theme.formBgColor};
  font-family: ${(props): string => props.theme.fontFamily};
  border-radius: 6px;
  box-shadow: ${(props): string => props.theme.formBoxShadow};
`

const FormWrapper: FC<FormProps> = ({
  children,
  onSubmit,
  customTheme,
  id,
  stepsLength,
  stepsTitles,
  hasRedirect,
  redirectUrl,
  timeToRedirect,
  redirectHeaderText,
  logoImg,
  redirectMainImg,
  redirectBgImg,
  sendDataToApi,
  apiUrl,
  transactionName,
  propertyNamesFromUrl,
  dataWithUserAgent,
}) => {
  const trasationIdValue =
    transactionName &&
    new URLSearchParams(window.location.search).get(transactionName)
  const intialValuesFromUrl = useMemo(
    () =>
      (propertyNamesFromUrl &&
        propertyNamesFromUrl.length > 0 &&
        getFieldsValuesFromUrl(propertyNamesFromUrl)) ||
      {},
    [propertyNamesFromUrl]
  )
  const [redirectUrlPath, setRedirectUrlPath] = useState(redirectUrl)
  const [initialValues, setInitialValues] = useState({
    ...intialValuesFromUrl,
    ...getFormValuesFromCache(id),
    // eslint-disable-next-line @typescript-eslint/camelcase
    trasaction_id: trasationIdValue,
  })
  const [currentStep, setCurrentStep] = useState(0)
  const [fieldsForSkip, setFieldsForSkip] = useState<string[]>([])
  const [errorFromApi, setErrorFromApi] = useState<boolean>(false)
  const shouldRedirect = !errorFromApi && hasRedirect
  const addFieldForSkip = (key: string): void =>
    setFieldsForSkip([...fieldsForSkip, key])

  const handleSubmit = async (
    values: FormValuesType,
    props: {
      resetForm: () => void
      setStatus: (status: string) => void
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): Promise<any> => {
    if (onSubmit) {
      onSubmit(values, props)
    }

    try {
      if (sendDataToApi && apiUrl) {
        const response = await handleSendDataToApi(
          values,
          apiUrl,
          id,
          fieldsForSkip,
          dataWithUserAgent
        )
        const {
          redirectUrl: urlFromApi,
          status: statusFromApi,
        } = await response.json()

        if (statusFromApi === false) {
          setErrorFromApi(true)
          return
        }

        urlFromApi && setRedirectUrlPath(urlFromApi)
      }

      props.resetForm()
      props.setStatus(formStatuses.submited)
    } catch (e) {
      props.setStatus(formStatuses.error)
    }
  }
  const handleReset = (): void => {
    resetFormValueCache(id)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setInitialValues({} as any)
  }

  const prevStep: Function = () => setCurrentStep(currentStep - 1)

  const nextStep: Function = () => setCurrentStep(currentStep + 1)

  return (
    <FormContext.Provider
      value={{
        id,
        apiUrl,
        stepsLength,
        currentStep,
        stepsTitleList: stepsTitles,
        nextStep,
        prevStep,
        fieldsForSkip,
        addFieldForSkip,
        errorFromApi,
      }}
    >
      <ThemeProvider customTheme={{ ...customTheme }}>
        <Formik
          enableReinitialize
          initialValues={initialValues}
          onSubmit={handleSubmit}
          onReset={handleReset}
        >
          {(props): ReactElement =>
            (shouldRedirect && props.status === formStatuses.submited && (
              <RedirectPage
                redirectUrl={redirectUrlPath}
                backgroundImage={redirectBgImg}
                logoImg={logoImg}
                headerText={redirectHeaderText}
                timeToRedirect={timeToRedirect}
                mainImg={redirectMainImg}
              />
            )) || (
              <StyledForm id={id}>
                {children}
              </StyledForm>
            )
          }
        </Formik>
      </ThemeProvider>
    </FormContext.Provider>
  )
}

export default FormWrapper
