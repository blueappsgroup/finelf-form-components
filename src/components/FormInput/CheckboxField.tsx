import React, { FC } from 'react'
import { Field } from 'formik'

import CheckboxBase from './CheckBox'
import { CheckboxFieldProps } from '../../types'
import { validateCheckbox } from './validate'

const CheckboxFormField: FC<CheckboxFieldProps> = ({
  id = 'textField',
  label,
  name = 'checkbox_field',
  required = false,
  onChange,
  children,
}) => (
  <Field
    required={required}
    id={id}
    name={name}
    component={CheckboxBase}
    label={label}
    type="checkbox"
    validate={validateCheckbox(required)}
    onChange={onChange}
    childrenBody={children}
  />
)

CheckboxFormField.defaultProps = {
  name: 'checkbox_field',
}

export default CheckboxFormField
