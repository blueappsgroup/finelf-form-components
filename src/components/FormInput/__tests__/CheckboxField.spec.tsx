import * as React from 'react'
import { render, fireEvent, act } from '@testing-library/react'

import Form from '../../Form'
import CheckboxField from '../CheckboxField'

describe('<CheckboxField />', () => {
  const onSubmit = jest.fn()
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const setupWrapper: (config: any) => any = ({
    formId = 'testForm',
    checkboxId = 'checkboxField',
    checkboxName = 'checkbox_field',
    required = false,
    ...rest
  }) => {
    const wrapper = render(
      <Form id={formId} onSubmit={onSubmit}>
        <CheckboxField
          id={checkboxId}
          name={checkboxName}
          required={required}
          {...rest}
        />
      </Form>
    )
    const { container } = wrapper

    const checkbox = container.querySelector(`[name="${checkboxName}"]`)

    return {
      checkbox,
      ...wrapper,
    }
  }

  it('matches snapshot', () => {
    const { checkbox } = setupWrapper({})

    expect(checkbox).toBeTruthy()
    expect(checkbox).toMatchSnapshot()
  })

  it('matches snapshot with label', () => {
    const wrapper = setupWrapper({ label: 'Test' })
    const { container } = wrapper

    expect(wrapper.getByText('Test')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('changes checkbox field', async () => {
    const { checkbox } = setupWrapper({})

    await act(async () => {
      fireEvent.click(checkbox)
    })

    expect(checkbox.checked).toBe(true)
  })

  it('checkbox field handleOnMouseOut save value to sessionStorage', async () => {
    const { checkbox } = setupWrapper({})

    await act(async () => {
      fireEvent.click(checkbox)
    })

    await act(async () => {
      fireEvent.mouseOut(checkbox)
    })

    expect(
      JSON.parse(global.window.sessionStorage.getItem('form-testForm'))
        .checkbox_field
    ).toBe(true)
    expect(checkbox.checked).toBe(true)
  })
})
