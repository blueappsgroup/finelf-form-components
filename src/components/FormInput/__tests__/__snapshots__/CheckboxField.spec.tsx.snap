// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxField /> matches snapshot 1`] = `
<input
  class="sc-iqHYGH eXUgkp"
  id="checkboxField"
  name="checkbox_field"
  type="checkbox"
  value="false"
/>
`;

exports[`<CheckboxField /> matches snapshot with label 1`] = `
<div>
  <form
    action="#"
    class="sc-iBPRYJ janqdG"
    id="testForm"
  >
    <div
      class="sc-fubCfw krrqIb"
    >
      <label
        class="sc-bqyKva cyvhyU"
      >
        <div
          class="sc-jrAGrp bNKtyb"
        >
          <input
            class="sc-iqHYGH eXUgkp"
            id="checkboxField"
            label="Test"
            name="checkbox_field"
            type="checkbox"
            value="false"
          />
          <div
            class="sc-crrsfI jYgHay"
          >
            <svg
              class="sc-kEjbxe fkpYrb"
              viewBox="0 0 24 24"
            >
              <polyline
                points="20 6 9 17 4 12"
              />
            </svg>
          </div>
        </div>
        <span
          class="sc-dQppl lbCxfG"
        >
          Test
        </span>
        <div
          class="sc-hBEYos eXyhIn"
        />
      </label>
      <span
        class="sc-pFZIQ lowQPr"
      />
    </div>
  </form>
</div>
`;
